src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass_LocalFunctions.cs

```diff
@@ -0,0 +1,164 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class AbstractFlowPass<TLocalState, TLocalFunctionState>
+    {
+        internal abstract class AbstractLocalFunctionState
+        {
+            /// <summary>
+            /// This is the part of the local function transfer function which
+            /// transfers knowledge additively. For example, in definite
+            /// assignment this would be captured state which is assigned by
+            /// the local function.  When a local function is called, this
+            /// state is <see cref="Meet(ref TLocalState, ref TLocalState)"/>
+            /// with the current state.
+            /// </summary>
+            public TLocalState State;
+
+            public AbstractLocalFunctionState(TLocalState unreachableState)
+            {
+                State = unreachableState;
+            }
+
+            public bool Visited = false;
+        }
+
+        protected abstract TLocalFunctionState CreateLocalFunctionState();
+
+        private readonly SmallDictionary<LocalFunctionSymbol, TLocalFunctionState> _localFuncVarUsages =
```

Comment:
Do we want to consider not allocating this unless we encounter a local function? It feels like, even in our local-function-happy codebase, this will be more unused than used.